### PR TITLE
Universal Viewer Dist and RIIIF and IIIF Manifest files

### PR DESCRIPTION
**Pull request by jenlindner**

### CODE REVIEW

**Comment by scossu** [comment]
Just to be clear, this is not a candidate for merging into 1.1.0. There is still some testing to be done.

**Comment by awead** [comment]
Note. Don't merge this until after 1.1.0 is released.

File: app/assets/stylesheets/viewer.scss
Comment (awead) [inline app/assets/stylesheets/viewer.scss]: can you remove the iframe if this is commented-out?

File: app/controllers/curation_concerns/generic_works_controller.rb
Comment (awead) [inline app/controllers/curation_concerns/generic_works_controller.rb]: This comment should be moved down so that it's right about `def show`. Also, could you explain what specific line or lines in the method you're overriding? I'm wondering that if you're changing the template or view code, you could change the template or just override the view code instead of the controller code.

File: Gemfile
Comment (awead) [inline Gemfile]: openseadragon is already included in line 17

File: app/models/file_set.rb
Comment (awead) [inline app/models/file_set.rb]: We shouldn't need this. It's in `config/initializers/hydra_works.rb`

File: app/models/solr_document.rb
Comment (awead) [inline app/models/solr_document.rb]: Hmm. I'm wondering why this is needed here. It's been overridden in `config/initializers/hydra_works.rb`

File: app/presenters/aic_work_show_presenter.rb
Comment (awead) [inline app/presenters/aic_work_show_presenter.rb]: This seems like a good candidate for an entry in the locales file!

File: app/presenters/aic_work_show_presenter.rb
Comment (awead) [inline app/presenters/aic_work_show_presenter.rb]: You've defined a new class here but it only has one subclass, so why the inheritance?

File: app/presenters/asset_presenter.rb
Comment (awead) [inline app/presenters/asset_presenter.rb]: AssetPresenter is the only class inheriting from AICWorkShowPresenter, so I would merge the two classes and put the methods in AICWorkShowPresenter in this class.

File: app/presenters/concerns/displays_image.rb
Comment (awead) [inline app/presenters/concerns/displays_image.rb]: This is hitting Fedora and we shouldn't do that in a presenter context. Run the solr query here and reutrn nil if the query is nil.

File: app/presenters/concerns/displays_image.rb
Comment (awead) [inline app/presenters/concerns/displays_image.rb]: You build a SolrDocument here:
``` ruby
hit = ActiveFedora::SolrService.query("id:#{id}")
doc = SolrDocument.new(hit)
```
Now, you can call `doc.height` and `doc.width`
see: https://github.com/projecthydra/active_fedora/blob/master/lib/active_fedora/solr_service.rb#L47

File: app/renderers/universal_viewer.rb
Comment (awead) [inline app/renderers/universal_viewer.rb:3]: Is UniversalViewer supposed to be a singleton? Why? I've heard singletons are "bad" just curious...

File: app/services/iiif_authorization_service.rb
Comment (awead) [inline app/services/iiif_authorization_service.rb]: Could put some class-level documentation in here? I'm guessing this is something that `Riif` uses? A service that takes a controller as a param caught me off-guard.

File: app/views/curation_concerns/base/show_asset.html.erb
Comment (awead) [inline app/views/curation_concerns/base/show_asset.html.erb:1]: Could add some comments to indicate what line(s) are being overridden?

File: config/secrets.yml
Comment (awead) [inline config/secrets.yml]: I would revert this change. It isn't germane to the PR, and the key should be indented.

File: app/services/manifest_service.rb
Comment (awead) [inline app/services/manifest_service.rb:59]: I think the better approach here is to create your own factory:
``` ruby
class PdfManifestFactory < IIIFManifest::ManifestFactory
  def add_media_sequences(manifest, media_info)
    [...]
end
```
But, isn't `manifest` above just `self.to_h`?

File: spec/spec_helper.rb
Comment (awead) [inline spec/spec_helper.rb]: use https://github.com/aic-collections/aicdams-lakeshore/blob/master/spec/support/lakeshore_testing.rb#L31 instead

**Review by awead — CHANGES_REQUESTED** [review]
@jenlindner looks good overall. It needs a rebase off of develop first, and then let's talk about the remaining bits.

File: app/presenters/aic_work_show_presenter.rb
Comment (jenlindner) [reply app/presenters/aic_work_show_presenter.rb]: @awead a brief yet clear entry! working on it now :-)

File: app/views/curation_concerns/base/show_asset.html.erb
Comment (jenlindner) [reply app/views/curation_concerns/base/show_asset.html.erb:1]: They're added, at line 15.

File: app/services/manifest_service.rb
Comment (jenlindner) [reply app/services/manifest_service.rb:59]: This is the main aspect I couldn't quite get, I tried in a variety of ways to make this work with its own factory. I don't think this is the worst way for the time being. And to_h is a method in the factory that adds an awful lot of behavior, isn't just the to_hash one might expect.

File: app/renderers/universal_viewer.rb
Comment (jenlindner) [reply app/renderers/universal_viewer.rb:3]: This is another thing I took pretty much without any questions from Hyku, so couldn't really answer. We could ask Trey or Justin if you'd like, though.

**Comment by jenlindner** [comment]
hi @awead, I have a question. There's caching happening on the rails side that prevents the viewer from loading a good portion of the time. I've experimented some this afternoon with different types of headers to force expiration with no luck, but a little more research turned up turbolinks as a likely culprit too. Does that sound plausible to you? I started looking into ways to leave them enabled for everything except the controller and action in question, but thought I'd ask for your feedback as well. Thanks.

**Comment by awead** [comment]
@jenlindner yes, that seems very likely. We could disable it altogether, if needed, but maybe there's a subtler way. Have you asked @jcoyne or others involved with Hyku?

**Comment by jenlindner** [comment]
Okay, I tested with turbolinks disabled site-wide and the problem disappears. Yay to finding the root cause.

So then, as to how to disable them only in this view: one way is to use a data attribute in link tags, but our code doesn't use an html link tag to load the viewer, and the viewer generates its own html. Another, which I think might be the way to go, is to put "data-no-turbolink" into the body tag, which I think might best be accomplished in a layout. (I'm getting information on disabling them from this post: http://blog.flightswithfriends.com/post/53943440505/how-to-disable-turbolinks-in-rails-4)

We use the Curation Concerns layout code on the show pages: 

https://github.com/projecthydra/curation_concerns/blob/master/app/controllers/concerns/curation_concerns/curation_concern_controller.rb#L9-L10

https://github.com/projecthydra/curation_concerns/blob/master/app/controllers/concerns/curation_concerns/themed_layout_controller.rb#L13-L21

This creates the body tag: https://github.com/projecthydra/curation_concerns/blob/master/app/views/layouts/curation_concerns.html.erb#L1

So, one option is to override this layout in LAKE to add the attribute to the body tag.

Another option might be to override the method with_themed_layout(view_name = nil) in LAKE to take a parameter for adding an attribute to the body tag. 

Do you have any recommendations, @awead @jcoyne? I haven't tested this yet, also, don't know for certain that the attribute in the body tag would solve the problem.

**Comment by awead** [comment]
@jenlindner I've seen `data-no-turbolink` before as an option. Does that fix the problem? If not, I'd say post to hydra-tech about this and see what Hyku is doing, otherwise I think we'd just have to nix turbolinks.

**Comment by jenlindner** [comment]
That does fix the problem, thanks @awead. I added it to a new layout, added the layout to the generic works controller and added a controller test that checks for both the new template and the layout in the show action.

**Review by awead — APPROVED** [review]

**Pull request merged**
